### PR TITLE
Update partial claim & balance retrieval

### PR DESCRIPTION
**Pull request by tintinthong**
- this both enables `claimRewards` to be used and it updates proof and balance retrieval to account for the new reward pool data structure 
- I DIDINT add the query for `/reward-balance/<payee>` that triggers the `known_claimed` state of USED proofs. Feels unusual to force a read action to update dbs. Perhaps, merge this first and think about ui later. 

- this is fork of [2415](https://github.com/cardstack/cardstack/pull/2415)
- relies on this https://github.com/cardstack/card-protocol-relay-service/pull/44

### CODE REVIEW

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 1]
```diff
@@ -327,41 +264,55 @@ export default class RewardPool {
...
+    }
+
+    let { rewardProgramId, payee, token, amount }: FullLeaf = this.decodeLeaf(leaf) as FullLeaf;
+
     if (!rewardProgramId) {
       throw new Error('rewardProgramId must be provided');
     }
-    if (!tokenAddress) {
-      throw new Error('tokenAddress must be provided');
+    if (!leaf) {
```
Comment (IanCal) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:280]: These checks duplicate some from above

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 2]
```diff
@@ -327,41 +264,55 @@ export default class RewardPool {
...
     }
-    if (!proof) {
-      throw new Error('proof must be provided');
+    if (!proofArray) {
+      throw new Error('proofArray must be provided');
+    }
+    if (!token) {
+      throw new Error('token must be provided');
+    }
+    if (!amount) {
```
Comment (IanCal) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:289]: Amount has been removed now

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 3]
```diff
@@ -327,41 +264,55 @@ export default class RewardPool {
...
-For the proof, the reward safe owner can only redeem ${unclaimedRewards} but user is asking for ${amount}`
-      );
+    if (rewardSafeOwner != payee) {
+      throw new Error('payee is not owner of the reward safe');
     }
 
+    if (!(await this.isValid(leaf, proofArray))) {
+      throw new Error('proof is not valid');
+    }
+    if (!(await this.isClaimed(leaf))) {
```
Comment (IanCal) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:310]: This is fails if the leaf has _not_ been claimed - I think it should be

    if (await this.isClaimed(leaf)) {

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 4]
```diff
@@ -327,41 +264,55 @@ export default class RewardPool {
...
+      throw new Error('reward has been claimed');
+    }
+    let from = contractOptions?.from ?? (await this.layer2Web3.eth.getAccounts())[0];
+    let weiAmount = new BN(amount);
+    let rewardPoolBalanceForRewardProgram = (await this.balance(rewardProgramId, token)).balance;
     if (weiAmount.gt(rewardPoolBalanceForRewardProgram)) {
```
Comment (IanCal) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:316]: This doesn't deal properly with partial claims - there we expect the amount may be lower. Perhaps just wrapping with an extra if and adding a note in the exception that a workaround is to accept a partial claim

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 5]
```diff
@@ -476,6 +422,61 @@ The reward program ${rewardProgramId} has balance equals ${fromWei(
...
     return o;
   }
+
+  private decodeLeaf(leaf: string): FullLeaf {
+    let outerObj = this.layer2Web3.eth.abi.decodeParameters(
+      [
+        { type: 'address', name: 'rewardProgramId' },
+        { type: 'uint256', name: 'paymentCycleNumber' },
+        { type: 'uint256', name: 'startBlock' },
+        { type: 'uint256', name: 'endBlock' },
```
Comment (IanCal) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:432]: A small one but we renamed this validFrom and validTo in the contract

File: packages/cardpay-sdk/sdk/reward-pool/base.ts [hunk 6]
```diff
@@ -14,28 +14,42 @@ import ERC677ABI from '../../contracts/abi/erc-677';
...
   rewardProgramId: string;
+  amount: BN;
+  leaf: string;
+}
+
+export interface Leaf {
+  rewardProgramId: string;
+  paymentCycleNumber: number;
+  startBlock: number;
+  endBlock: number;
```
Comment (IanCal) [inline packages/cardpay-sdk/sdk/reward-pool/base.ts:35]: validFrom/validTo as well here I think (as in the other comment)

**Review by IanCal — CHANGES_REQUESTED** [review]
Looking good, agree on not using the reward balances part of tally. Mostly just some small comments.

File: packages/cardpay-sdk/sdk/reward-pool/base.ts (hunk 3 above)
Comment (tintinthong) [reply packages/cardpay-sdk/sdk/reward-pool/base.ts:310]: Right yea. This was a check I added after I tested. Thanks good spot!

File: packages/cardpay-sdk/sdk/reward-pool/base.ts (hunk 2 above)
Comment (tintinthong) [reply packages/cardpay-sdk/sdk/reward-pool/base.ts:283]: I think this is relevant for checking how much reward is being taken out and comparing that to whats avaialbe in reward pool/ whats gas costing

File: packages/cardpay-sdk/sdk/reward-pool/base.ts (hunk 4 above)
Comment (tintinthong) [reply packages/cardpay-sdk/sdk/reward-pool/base.ts:310]: I made a fix here

File: packages/cardpay-sdk/sdk/reward-pool/base.ts (hunk 2 above)
Comment (IanCal) [reply packages/cardpay-sdk/sdk/reward-pool/base.ts:283]: Sorry, I missed the section taking the amount from the leaf and thought it was left over from when the amount needed to be specified by the user.

**Review by IanCal — APPROVED** [review]
LGTM, not run locally but code changes make sense to me.

**Pull request merged**